src/typechecker.erl

```diff
@@ -3200,13 +3310,33 @@ refine(OrigTy, Ty, TEnv) ->
         RefTy  -> RefTy
     end.
 
+get_record_fields_types(Name, Anno, TEnv) ->
+    RecordFields = get_maybe_remote_record_fields(Name, Anno, TEnv),
+    [Type || ?typed_record_field(_, Type) <- RecordFields].
+
+expand_record(Name, Anno, TEnv) ->
+    type_record(Name, get_record_fields_types(Name, Anno, TEnv)).
+
 %% May throw no_refinement.
 refine_ty(_Ty, ?type(none), _TEnv) ->
     %% PatTy none() means the pattern can't be used for refinement,
     %% because there is imprecision.
     throw(no_refinement);
 refine_ty(?type(T, A), ?type(T, A), _) ->
     type(none);
+refine_ty(?type(record, [{atom, _, Name}]), ?type(record, [{atom, _, Name}]), _) ->
+    type(none);
+refine_ty(?type(record, [{atom, Anno, Name}]), Refined = ?type(record, [{atom, _, Name} | _]), TEnv) ->
+    refine_ty(expand_record(Name, Anno, TEnv), Refined, TEnv);
+refine_ty(Refined = ?type(record, [{atom, _, Name} | _]), ?type(record, [{atom, Anno, Name}]), TEnv) ->
+    refine_ty(Refined, expand_record(Name, Anno, TEnv), TEnv);
```

Comment:
Shouldn't this be none()? If not, why not?

Reply:
So at this point, we might be the first function clause and the record type hasn't been expanded yet. An example:

```erlang
-record(r, {f :: integer() | undefined}).

-spec test(#r{}) -> integer().
test(#r{f = undefined}) -> 0;
test(#r{f = I}) -> I.
```

At this point, the argument type is `{type, _, record, [{atom, _, r}]}` but the pattern type would be `{type, _, record, [{atom, _, r}, {atom, _, undefined}]}`.

Reply:
If the first argument to `refine_ty` is not expanded, I agree, it has to be expanded in order to refine it. (Lines 3329-3330)

If the second argument to `refine_ty` is not expanded, I don't agree. This means that the result is none() and the record type doesn't need to be expanded to compute that. Perhaps this second clause isn't even needed...? (Lines 3331-3332)

Reply:
Ah right, that makes sense. First argument is always the least refined right? I will change this.

Reply:
> First argument is always the least refined right?

Well, this is like "minus" for types. If the first argument is a subtype of the second, then the result should be none().

This function is described as a helper for `type_diff`. (I think it would be a good idea to rename it to e.g. `type_diff_help` or `do_type_diff` or so, but we'd better make such refactorings as separate PRs.)